### PR TITLE
Fix: leading comments added from previous node (fixes #5531)

### LINKED ISSUE #5531: Leading comments incorrectly include comments inside previous token
I recently contributed a rule that ended up having an issue regarding leading comments. The issue for the rule and the fix is [here](https://github.com/eslint/eslint/issues/5480), but in looking at it, I started wondering if maybe this should be a fix in `sourceCode.getComments` rather than having to be guarded against in each individual rule.

I'm currently using ESLint v2.3.0, and the problem seems to be that comments inside a previous object literal or block get added to the following token's leading comments array when there are no other expressions between the comment and the token. Example:

```
function foo() {
  var bar = {
    // comment
  };

  return bar;
}
```
Running `sourceCode.getComments` on the `return` expression's node will include the comment inside the previous object literal, despite it actually being part of the previous token.

I'm assuming this is a bug - it definitely wasn't what I was expecting to happen - can someone more knowledgable confirm?

In looking at things, it looks like maybe the change needs to happen in Espree...Happy to make a PR if this is accepted.

### PR DESCRIPTION
**Pull request by kaicataldo**

### CODE REVIEW

File: tests/lib/rules/newline-before-return.js [hunk 1]
```diff
@@ -71,6 +70,7 @@ ruleTester.run("newline-before-return", rule, {
         "function a() {\nswitch (b) {\ncase 'b':\nreturn;\n}\n}",
         "function a() {\nswitch (b) {\ncase 'b': {\nreturn;\n}\n}\n}",
         "function a() {\n//comment\nreturn b;\n}",
+        "function a() {\nvar b = {\n//comment\n};\n\nreturn;\n}",
```
Comment (kaicataldo) [inline tests/lib/rules/newline-before-return.js]: Moved to be with other comment tests

File: package.json [hunk 2]
```diff
@@ -42,7 +42,7 @@
     "doctrine": "^1.2.0",
     "es6-map": "^0.1.3",
     "escope": "^3.6.0",
-    "espree": "^3.1.2",
+    "espree": "git+https://github.com/eslint/espree.git",
```
Comment (kaicataldo) [inline package.json]: Will change to latest version of espree when it's released, but even running `npm install espree@next` only installed the latest tagged version, not the latest version from the repo. Wanted to make the PR with this version to get feedback.

File: lib/rules/newline-before-return.js [hunk 3]
```diff
@@ -75,23 +75,6 @@ module.exports = function(context) {
         }
 
         comments.forEach(function(comment) {
-            // https://github.com/eslint/eslint/pull/5512
```
Comment (kaicataldo) [inline lib/rules/newline-before-return.js]: Removed guard that alerted me to this issue in the first place (while keeping the test), as its now unnecessary (yay)

**Comment by nzakas** [comment]
Lgtm. I'll try to get a new Espree version pushed today

**Comment by kaicataldo** [comment]
Thanks - updated this PR with the latest version of espree

**Comment by ilyavolodin** [comment]
Thanks!

**Pull request merged**